File: frontend/src/types/role-management.ts
/**
 * 角色管理的 TypeScript 类型定义
 */

// ==================== 角色相关类型 ====================

/** 角色信息 */
export interface Role {
  id: number
  name: string
  description: string | null
  status: 'active' | 'inactive'
  created_at: string
  updated_at: string
}

/** 角色统计信息 */
export interface RoleStats {
  total_roles: number
  active_roles: number
  inactive_roles: number
  roles: Array<Role & { user_count: number }>
}

/** 权限矩阵（模块 → 操作 → 是否授权） */
export interface PermissionMatrix {
  [module: string]: {
    [action: string]: boolean
  }
}

/** 带权限的角色信息 */
export interface RoleWithPermissions extends Role {
  permissions: PermissionMatrix
}

// ==================== API 请求类型 ====================

/** 角色创建请求 */
export interface RoleCreateRequest {
  name: string
  description?: string
  status?: 'active' | 'inactive'
}

/** 角色更新请求 */
export interface RoleUpdateRequest {
  name?: string
  description?: string
  status?: 'active' | 'inactive'
}

/** 角色权限更新请求 */
export interface RolePermissionsUpdateRequest {
  permissions: PermissionMatrix
}

// ==================== API 响应类型 ====================

/** 角色列表响应 */
export interface RoleListResponse {
  data: Role[]
  meta: {
    total: number
  }
}

/** 角色详情响应 */
export interface RoleResponse {
  data: Role
}

/** 角色权限响应 */
export interface RolePermissionsResponse {
  data: PermissionMatrix
}

/** 角色统计响应 */
export interface RoleStatsResponse {
  data: RoleStats
}

/** 通用操作响应 */
export interface RoleActionResponse {
  data: {
    success: boolean
    [key: string]: any
  }
  message?: string
}

// ==================== Store 状态类型 ====================

/** 角色管理状态 */
export interface RoleManagementState {
  roles: Role[]
  currentRole: RoleWithPermissions | null
  stats: RoleStats | null
  isLoading: boolean
  isSaving: boolean
  error: string | null
}

// ==================== 表单类型 ====================

/** 角色表单数据 */
export interface RoleFormData {
  name: string
  description: string
  status: 'active' | 'inactive'
}

/** 角色权限配置数据 */
export interface RolePermissionFormData {
  roleId: number
  permissions: PermissionMatrix
}

File: backend/app/schemas/role_management.py
"""
角色管理 Pydantic Schema - 请求/响应验证
"""
import re
from pydantic import BaseModel, Field, field_validator
from typing import Dict, List, Optional
from datetime import datetime


class RoleCreateRequest(BaseModel):
    """角色创建请求"""
    name: str = Field(..., description="角色名称", min_length=1, max_length=50)
    description: Optional[str] = Field(None, description="角色描述", max_length=500)
    status: Optional[str] = Field("active", description="状态", pattern="^(active|inactive)$")
    
    @field_validator('name')
    @classmethod
    def validate_name(cls, v: str) -> str:
        # 检查名称格式：以小写字母开头，只能包含字母、数字和下划线
        if not re.match(r'^[a-z][a-z0-9_]{2,49}$', v):
            raise ValueError('角色名称必须以小写字母开头，只能包含字母、数字和下划线，长度 3-50 个字符')
        
        # 检查是否为系统默认角色
        valid_names = ['admin', 'manager', 'specialist', 'sales']
        if v.lower() in valid_names:
            raise ValueError(f'无法创建系统默认角色：{v.lower()}。系统默认角色已存在')
        
        return v


class RoleUpdateRequest(BaseModel):
    """角色更新请求"""
    name: Optional[str] = Field(None, description="角色名称", min_length=1, max_length=50)
    description: Optional[str] = Field(None, description="角色描述", max_length=500)
    status: Optional[str] = Field(None, description="状态", pattern="^(active|inactive)$")


class RolePermissionsUpdateRequest(BaseModel):
    """角色权限更新请求"""
    permissions: Dict[str, Dict[str, bool]] = Field(
        ...,
        description="权限字典 {module: {action: granted}}"
    )


class RoleResponse(BaseModel):
    """角色响应"""
    id: int
    name: str
    description: Optional[str] = None
    status: str
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class RoleWithPermissionsResponse(BaseModel):
    """带权限的角色响应"""
    id: int
    name: str
    description: Optional[str] = None
    status: str
    permissions: Dict[str, Dict[str, bool]]
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class RoleListResponse(BaseModel):
    """角色列表响应"""
    data: List[RoleResponse]
    meta: Dict[str, int]


class RoleStatsResponse(BaseModel):
    """角色统计响应"""
    total_roles: int
    active_roles: int
    inactive_roles: int
    roles: List[Dict]
